### ISSUE TITLE
Accessing the contentSize of a PDFPageLayout

### ISSUE THREAD

**Issue by omiz** [techprimate/TPPDF#154]
The variable `contentSize` is unavailable after updating to version `1.5.4'

is it coming back or does the page area have a new calculation for the content area?

**Comment by techprimate-phil** [comment]
This value got removed in `1.5.0` as we had changed the generator to actually track the margin, instead of the layout, due to sections

**Comment by omiz** [comment]
Unfortunately sections do not work with my input. I need to calculate the needed size manually for some of the items that I have.

so is it possible somehow to get the content size before adding the content or generating the document

is other words would the old implementation work

```swift
var contentSize: CGSize {
    
    let width = size.width - margin.left - margin.right
    let height = size.height - margin.top - space.header - space.footer - margin.bottom
    
    return CGSize(width: width, height: height)
}
```

**Comment by techprimate-phil** [comment]
Take a look at `PDFCalculations.calculateAvailableFrame:102`. Most of the objects use this method to calculate the available width, which is the content size.

I do not think the actual calculations got changed, but instead the structure of tracking layout changes (moving the layout from PDFDocument to PDFGenerator, so PDFDocuments don't get changed by the generator).

Therefore, yes this should work in theory

**Comment by omiz** [comment]
I'm struggling with accessing any of the classes used because of the `internal` accessor

A suggestion would be:
as I see from the implementation as of version 1.5.4 we can still access this through these extensions

```swift
extension TPPDF.PDFPageLayout {
    
    var contentSize: CGSize {
        
        let width = size.width - margin.left - margin.right
        let height = size.height - margin.top - space.header - space.footer - margin.bottom
        
        return CGSize(width: width, height: height)
    }
}

extension TPPDF.PDFDocument {

    var contentSize: CGSize {
        return layout.contentSize
    }
    
}
```

This is a hot fix and not meant to be used as a final solution.

**Comment by techprimate-phil** [comment]
About the `internal`: too keep the API clean, most internal classes are hidden but can be changed to public accessors if necessary.

The main reason why the `contentSize` got removed from the `PDFPageLayout` is, that this class should be used for setting up the document, and not actually performing any calculations.
Also since we do have `PDFMarginObject` in-use for groups (e.g. to set a different margin for master groups) the `margin.top`, `margin.left` etc. are not the only margins in the framework.

Your solution might be a hotfix for your use-case, but as it does not use all parameters, I would prefer to not put it back in as an extension. I hope you understand my position.

As of your problem:

```swift
let width = size.width - margin.left - margin.right
let height = size.height - margin.top - space.header - space.footer - margin.bottom
```
All of these values are provided by you to the framework, so why would you need it back in `PDFDocument`?

**Comment by omiz** [comment]
My main issue is not being able to use the provided implementation of Sections.

the data that I have is a little bit random and it crash the layout. therefor I need to make that calculation manually

I've already mentioned that in this issue [here](https://github.com/techprimate/TPPDF/issues/73).

**Comment by philprime** [comment]
Hi there, I just commented on your referenced issue.
As I still believe this is a hackish work-around for your specific use case and not the actual purpose of this class nor a solution for the issue, I will close this now.
Hope you understand, and we'll keep working to fix the section issue in #73
